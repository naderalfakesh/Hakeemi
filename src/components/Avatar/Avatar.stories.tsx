import React, { Fragment } from 'react';
import { View, StyleSheet, Image } from 'react-native';
import { storiesOf } from '@storybook/react-native';
import Avatar, { Props } from './Avatar';

const Container = (storyFn: any) => (
  <View style={styles.container}>{storyFn()}</View>
);

const ImageDemo = () => (
  <Image
    source={{
      uri: 'https://i.ibb.co/4WGLRjs/person.png',
      width: 50,
      height: 50,
    }}
  />
);

const AvatarDemo = ({ theme }: { theme: Props['theme'] }) => (
  <Fragment>
    <Avatar size={50} theme={theme} shape="straight">
      <ImageDemo />
    </Avatar>
    <Avatar size={50} theme={theme} shape="rounded">
      <ImageDemo />
    </Avatar>
    <Avatar size={50} theme={theme} shape="circular">
      <ImageDemo />
    </Avatar>
  </Fragment>
);

storiesOf('Avatar', module)
  .addDecorator(Container)
  .add('primary', () => <AvatarDemo theme="primary" />)
  .add('secondary', () => <AvatarDemo theme="secondary" />);

const styles = StyleSheet.create({
  container: {
    flexGrow: 1,
    padding: 8,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-evenly',
  },
});
