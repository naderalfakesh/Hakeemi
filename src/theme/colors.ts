export const palette = {
  primary: {
    base: 'rgba(0, 188, 212, 1)',
    light: 'rgba(214, 250, 255, 1)',
    dark: 'rgba(144, 180, 206, 1)',
  },
  secondary: {
    base: 'rgba(237, 243, 247, 1)',
  },
  danger: {
    base: 'rgba(255, 87, 34, 1)',
  },
  grey: {
    base: 'rgba(240, 240, 240, 1)',
    light: 'rgba(251, 251, 251, 1)',
    dark: 'rgba(144, 144, 144, 1)',
  },
  white: {
    base: 'rgba(255, 255, 255, 1)',
  },
  black: {
    base: 'rgba(0, 0, 0, 1)',
  },
};

export default {
  primary: palette.primary.base,
  secondary: palette.secondary.base,
  danger: palette.danger.base,
  light: palette.grey.light,
  dark: palette.grey.dark,
  white: palette.white.base,
  black: palette.black.base,
};
