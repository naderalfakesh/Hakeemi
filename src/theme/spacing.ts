const spacing = {
  /** Space 2 */
  tiny: 2,
  /** Space 4 */
  mini1: 4,
  /** Space 6 */
  mini2: 6,
  /** Space 8 */
  mini3: 8,
  /** Space 10 */
  small1: 10,
  /** Space 12 */
  small2: 12,
  /** Space 16 */
  small3: 16,
  /** Space 18 */
  small4: 18,
  /** Space 20 */
  medium1: 20,
  /** Space 24 */
  medium2: 24,
  /** Space 32 */
  medium3: 32,
  /** Space 40 */
  big1: 40,
  /** Space 50 */
  big2: 50,
  /** Space 68 */
  big3: 68,
  /** Space 76 */
  huge1: 76,
  /** Space 88 */
  huge2: 88,
  /** Space 92 */
  huge3: 92,
};
export default spacing;
