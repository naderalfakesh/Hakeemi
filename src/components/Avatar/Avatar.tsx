import React, { FC } from 'react';
import { Pressable, StyleSheet, ViewStyle } from 'react-native';
import { palette } from '../../theme/colors';

export interface Props {
  style?: ViewStyle;
  size: number;
  theme?: 'primary' | 'secondary';
  shape?: 'rounded' | 'circular' | 'straight';
  roundness?: number;
  onPress?: () => void;
}

const Avatar: FC<Props> = ({
  style,
  size = 24,
  theme = 'primary',
  shape = 'straight',
  roundness = 0.25,
  onPress,
  children,
}) => {
  const borderRadius =
    shape === 'rounded'
      ? roundness * size
      : shape === 'circular'
      ? 0.5 * size
      : 0;

  const styles = StyleSheet.create({
    container: {
      width: size,
      height: size,
      borderRadius: borderRadius,
      overflow: 'hidden',
      backgroundColor:
        theme === 'primary' ? palette.primary.light : palette.grey.base,
    },
    pressed: {
      backgroundColor:
        theme === 'primary' ? palette.primary.dark : palette.grey.dark,
    },
  });

  return (
    <Pressable
      testID="avatar"
      style={({ pressed }) => [
        styles.container,
        pressed ? styles.pressed : undefined,
        style,
      ]}
      onPress={onPress}>
      {children}
    </Pressable>
  );
};

export default Avatar;
