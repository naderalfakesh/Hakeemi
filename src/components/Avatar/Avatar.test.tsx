import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import Avatar from './Avatar';
import { palette } from '../../theme/colors';
import { Image, StyleSheet } from 'react-native';

const SIZE = 50;
const ImageDemo = () => (
  <Image
    source={{
      uri: 'https://i.ibb.co/4WGLRjs/person.png',
      width: SIZE,
      height: SIZE,
    }}
  />
);

describe('Avatar component test', () => {
  it('renders correctly with default params', () => {
    const { toJSON } = render(
      <Avatar size={SIZE}>
        <ImageDemo />
      </Avatar>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders secondary circular avatar', () => {
    const { getByTestId } = render(
      <Avatar size={SIZE} theme="secondary" shape="circular">
        <ImageDemo />
      </Avatar>,
    );
    const avatar = getByTestId('avatar');
    const style = StyleSheet.flatten(avatar.props.style);
    expect(style.backgroundColor).toBe(palette.grey.base);
    expect(style.borderRadius).toBe(0.5 * SIZE);
  });

  it('renders rounded avatar with custom edge radius', () => {
    const ROUND_SIZE = 0.2;
    const { getByTestId } = render(
      <Avatar size={SIZE} shape="rounded" roundness={ROUND_SIZE}>
        <ImageDemo />
      </Avatar>,
    );
    const avatar = getByTestId('avatar');
    const style = StyleSheet.flatten(avatar.props.style);
    expect(style.borderRadius).toBe(ROUND_SIZE * SIZE);
  });

  it('overrides default style with style prop', () => {
    const CUSTOM_PADDING = 32;
    const { getByTestId } = render(
      <Avatar size={SIZE} style={{ padding: CUSTOM_PADDING }}>
        <ImageDemo />
      </Avatar>,
    );
    const avatar = getByTestId('avatar');
    const style = StyleSheet.flatten(avatar.props.style);
    expect(style.padding).toBe(CUSTOM_PADDING);
  });

  it('invokes a function when pressed', () => {
    const mockFn = jest.fn();
    const { getByTestId } = render(
      <Avatar size={SIZE} onPress={mockFn}>
        <ImageDemo />
      </Avatar>,
    );
    const avatar = getByTestId('avatar');
    fireEvent.press(avatar);
    expect(mockFn).toBeCalled();
  });
});
