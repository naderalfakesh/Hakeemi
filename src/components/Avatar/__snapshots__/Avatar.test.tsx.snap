// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component test renders correctly with default params 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "rgba(214, 250, 255, 1)",
        "borderRadius": 0,
        "height": 50,
        "overflow": "hidden",
        "width": 50,
      },
      undefined,
      undefined,
    ]
  }
  testID="avatar"
>
  <Image
    source={
      Object {
        "height": 50,
        "uri": "https://i.ibb.co/4WGLRjs/person.png",
        "width": 50,
      }
    }
  />
</View>
`;
